src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+            // and start node is parented by the parent of the end node (because binary operators are left associative)
+            // if this is the case - expand the selection to the entire parent of end node (in this case it will be [1 + 2 + 3] + 4)
+            const startParent = skipParentheses(start.parent);
+            const endParent = skipParentheses(end.parent);
+            if (isBinaryExpression(startParent) && isBinaryExpression(endParent) && isNodeDescendantOf(startParent, endParent)) {
+                start = end = endParent;
+            }
+            else {
+                // start and end nodes belong to different subtrees
+                return { errors: [createFileDiagnostic(sourceFile, span.start, length, Messages.CannotExtractFunction)] };
+            }
+        }
+        if (start !== end) {
+            // start and end should be statements and parent should be either block or a source file
+            if (!isBlockLike(start.parent)) {
+                return { errors: [createFileDiagnostic(sourceFile, span.start, length, Messages.CannotExtractFunction)] };
+            }
+            const statements: Statement[] = [];
+            for (const n of (<BlockLike>start.parent).statements) {
+                if (n === start || statements.length) {
+                    const errors = checkNode(n);
+                    if (errors) {
+                        return { errors };
+                    }
+                    statements.push(n);
+                }
+                if (n === end) {
+                    break;
+                }
+            }
+            return { targetRange: { range: statements, facts } };
+        }
+        else {
+            const errors = checkRootNode(start) || checkNode(start);
+            if (errors) {
+                return { errors };
+            }
+            const range = isStatement(start)
+                ? [start]
+                : <Expression>start;
+            return { targetRange: { range, facts } };
+        }
+
+        function checkRootNode(n: Node): Diagnostic[] | undefined {
+            if (isIdentifier(n)) {
+                return [createDiagnosticForNode(n, Messages.InsufficientSelection)];
+            }
+            return undefined;
+        }
+
+        // Verifies whether we can actually extract this node or not.
+        function checkNode(n: Node): Diagnostic[] | undefined {
+            const enum PermittedJumps {
+                None = 0,
+                Break = 1 << 0,
+                Continue = 1 << 1,
+                Return = 1 << 2
+            }
+            if (!isStatement(n) && !isExpression(n)) {
+                return [createDiagnosticForNode(n, Messages.StatementOrExpressionExpected)];
+            }
+
+            let errors: Diagnostic[];
+            let permittedJumps = PermittedJumps.Return;
+            let seenLabels: string[];
+
+            visit(n);
+
+            return errors;
+
+            function visit(n: Node) {

```

Comment:
Renamed. All recursive references are through `forEachChild` so this is effectively the same as our normal treewalking